File: Problem_1.py
# Time Complexity: O(n)
# Space Complexity: O(n)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        ugly = [1]
        i2, i3, i5 = 0, 0, 0
        while (len(ugly) < n):
            next_ugly = (min((ugly[i2] * 2), (ugly[i3] * 3), (ugly[i5] * 5)))
            ugly.append(next_ugly)
            if next_ugly == (ugly[i2] * 2):
                i2 += 1
            if next_ugly == (ugly[i3] * 3):
                i3 += 1
            if next_ugly == (ugly[i5] * 5):
                i5 += 1
        return ugly[-1]

File: Problem1.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class Solution {
    public int nthUglyNumber(int n) {
        int p2 = 0, p3 = 0, p5 = 0;
        int n2 = 2, n3 = 3, n5 = 5;
        int[] arr = new int[n];
        arr[0] = 1;
        for(int i = 1; i < n; i++){
            int mini = Math.min(n2,Math.min(n3,n5));
            if(mini == n2){
                arr[i] = n2;
                p2++;
                n2 = arr[p2]*2;
            }
            if(mini == n3){
                arr[i] = n3;
                p3++;
                n3 = arr[p3]*3;
            }
            if(mini == n5){
                arr[i] = n5;
                p5++;
                n5 = arr[p5]*5;
            }
        }
        return arr[n-1];
    }
}